src/animation/KeyframeTrack.js

```diff
@@ -26,61 +35,124 @@ THREE.KeyframeTrack.prototype = {
 
 	constructor: THREE.KeyframeTrack,
 
-	getAt: function( time ) {
+	TimeBufferType: Float64Array,
+	ValueBufferType: Float64Array,
 
+	DefaultInterpolation: THREE.InterpolateLinear,
 
-		// this can not go higher than this.keys.length.
-		while( ( this.lastIndex < this.keys.length ) && ( time >= this.keys[this.lastIndex].time ) ) {
-			this.lastIndex ++;
-		};
+	InterpolantFactoryMethodDiscrete: function( result ) {
 
-		// this can not go lower than 0.
-		while( ( this.lastIndex > 0 ) && ( time < this.keys[this.lastIndex - 1].time ) ) {
-			this.lastIndex --;
-		}
+		return new THREE.DiscreteInterpolant(
+				this.times, this.values, this.getValueSize(), result );
+
+	},
+
+	InterpolantFactoryMethodLinear: function( result ) {
+
+		return new THREE.LinearInterpolant(
+				this.times, this.values, this.getValueSize(), result );
+
+	},
+
+	InterpolantFactoryMethodSmooth: function( result ) {
+
+		return new THREE.SmoothInterpolant(
+				this.times, this.values, this.getValueSize(), result );
+
+	},
+
+	setInterpolation: function( interpolation ) {
+
+		var factoryMethod = undefined;
 
-		if ( this.lastIndex >= this.keys.length ) {
+		switch ( interpolation ) {
 
-			this.setResult( this.keys[ this.keys.length - 1 ].value );
+			case THREE.InterpolateDiscrete:
 
-			return this.result;
+				factoryMethod = this.InterpolantFactoryMethodDiscrete;
+
+				break;
+
+			case THREE.InterpolateLinear:
+
+				factoryMethod = this.InterpolantFactoryMethodLinear;
+
+				break;
+
+			case THREE.InterpolateSmooth:
+
+				factoryMethod = this.InterpolantFactoryMethodSmooth;
+
+				break;
 
 		}
 
-		if ( this.lastIndex === 0 ) {
+		if ( factoryMethod === undefined ) {
 
-			this.setResult( this.keys[ 0 ].value );
+			var message = "unsupported interpolation for " +
+					this.ValueTypeName + " keyframe track named " + this.name;
 
-			return this.result;
+			if ( this.createInterpolant === undefined ) {
+
+				// fall back to default, unless the default itself is messed up
+				if ( interpolation !== this.DefaultInterpolation ) {
+
+					this.setInterpolation( this.DefaultInterpolation );
+
+				} else {
+
+					throw new Error( message ); // fatal, in this case
+
+				}
+
+			}
+
+			console.warn( message );
+			return;
 
 		}
 
-		var prevKey = this.keys[ this.lastIndex - 1 ];
-		this.setResult( prevKey.value );
+		this.createInterpolant = factoryMethod;
+
+	},
+
+	getInterpolation: function() {
+
+		switch ( this.createInterpolant ) {
```

Comment:
No.

I completely misread you here. Now I get what "THREE define" means... I was thinking of the `THREE[ variable ]` stuff, but it's not what you meant...

> Why don't we just remember the THREE define of the interpolant that was set?

Could do. It would be redundant information and it's not needed in time critical code. There are many tracks. That's why I wrote it the way I did.

> Or alternatively, as the Interpolant what its type is?

No interpolant to ask. Just a factory method that holds it... Could put the constants on these methods, and I guess it's more elegant, however, I found it too offbeat.

> You mean recomputing this value via a switch statement is faster and also less code than returning a stored variable? I think we are misunderstanding each other.

No. It's slower and therefore not redundant. Used by `.toJSON` (a new feature I forgot to mention in the PR description BTW).